### PR TITLE
fix(resources): add resource limits/requests for operator deployment

### LINKED ISSUE #405: Define resource requests/limits for the operator deployment
The operator pod must have resource requests and should also have resource limits: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/.

We can give these values a try and see if they work for us:
https://github.com/operator-framework/operator-sdk/blob/280335593c76c1fb14ce96100938f140ddb42659/testdata/go/v3/memcached-operator/config/manager/manager.yaml#L50-L58

### PR DESCRIPTION
**Pull request by tthvo**
Fixes #405 

Added resource limits/requests for operator deployment.

### CODE REVIEW

**Comment by tthvo** [comment]
Tests passed and it seems operator deployment appears stable.

```
NAME                                                              READY   STATUS      RESTARTS   AGE
cryostat-operator-controller-manager-5f7f467b56-fdrtl             1/1     Running     0          36m
f1a07841fb2e41a3e1ea2964f5368d1624fc76fd8fb6b7522f1d82f8c0nbbzd   0/1     Completed   0          37m
quay-io-thvo-cryostat-operator-bundle-2-0-0-dev                   1/1     Running     0          37m
```

Not sure if we have any other criteria for choosing values for resource limits/requests? I used values as suggested from [here](https://github.com/operator-framework/operator-sdk/blob/280335593c76c1fb14ce96100938f140ddb42659/testdata/go/v3/memcached-operator/config/manager/manager.yaml#L50-L58).

File: bundle/manifests/cryostat-operator.clusterserviceversion.yaml [hunk 1]
```diff
@@ -509,7 +509,13 @@ spec:
                     port: 8081
                   initialDelaySeconds: 5
                   periodSeconds: 10
-                resources: {}
+                resources:
+                  limits:
+                    cpu: 500m
+                    memory: 123Mi
```
Comment (andrewazores) [inline bundle/manifests/cryostat-operator.clusterserviceversion.yaml:515]: 123? Kind of an odd amount. 128?

If you actually run the Operator and use `oc describe` on its Pod, you should be able to get some values for the amount of resources that our Operator actually uses in practice. Have you checked that yet?

Comment (tthvo) [reply bundle/manifests/cryostat-operator.clusterserviceversion.yaml:515]: > 123? Kind of an odd amount. 128?

Weird :D Guess I typed 3 and mistook it for 8 since its pretty much "half" of 8 somehow...

> If you actually run the Operator and use oc describe on its Pod, you should be able to get some values for the amount of resources that our Operator actually uses in practice. Have you checked that yet?

Right, I run it and it somehow did not show any resource usage. I got the logs [here](https://gist.github.com/tthvo/d7ffeedefa3245e397e2eecb30360b6b).

I also tried running `oc adm top pod cryostat-operator-controller-manager-8588cb98b8-zrgwl` but give:
```
error: Metrics API not available
```
I also try seeing the deployment configuration for the operator. It seems to not include any cpu or memory requests/limits. Maybe I didnt understand it correctly or somehow its got removed (already run `make bundle` after changes)?

```
Name:                   cryostat-operator-controller-manager
Namespace:              default
CreationTimestamp:      Tue, 05 Jul 2022 12:26:14 -0400
Labels:                 olm.deployment-spec-hash=576685dcf4
                        olm.owner=cryostat-operator.v2.2.0-dev
                        olm.owner.kind=ClusterServiceVersion
                        olm.owner.namespace=default
                        operators.coreos.com/cryostat-operator.default=
Annotations:            deployment.kubernetes.io/revision: 1
Selector:               control-plane=controller-manager
Replicas:               1 desired | 1 updated | 1 total | 1 available | 0 unavailable
StrategyType:           RollingUpdate
MinReadySeconds:        0
RollingUpdateStrategy:  25% max unavailable, 25% max surge
Pod Template:
  Labels:           control-plane=controller-manager
  Annotations:      alm-examples:
                      [
                        {
                          "apiVersion": "operator.cryostat.io/v1beta1",
                          "kind": "Cryostat",
                          "metadata": {
                            "name": "cryostat-sample"
                          },
                          "spec": {
                            "enableCertManager": true,
                            "eventTemplates": [],
                            "minimal": false,
                            "reportOptions": {
                              "replicas": 0
                            },
                            "storageOptions": {
                              "pvc": {
                                "annotations": {},
                                "labels": {},
                                "spec": {}
                              }
                            },
                            "trustedCertSecrets": []
                          }
                        },
                        {
                          "apiVersion": "operator.cryostat.io/v1beta1",
                          "kind": "FlightRecorder",
                          "metadata": {
                            "name": "example-flightrecorder"
                          },
                          "spec": {
                            "recordingSelector": {
                              "matchLabels": {
                                "operator.cryostat.io/flightrecorder": "example-flightrecorder"
                              }
                            }
                          }
                        },
                        {
                          "apiVersion": "operator.cryostat.io/v1beta1",
                          "kind": "Recording",
                          "metadata": {
                            "name": "example-recording"
                          },
                          "spec": {
                            "archive": true,
                            "duration": "30s",
                            "eventOptions": [
                              "template=ALL"
                            ],
                            "flightRecorder": {
                              "name": "example-flightrecorder"
                            },
                            "name": "example-recording"
                          }
                        }
                      ]
                    capabilities: Basic Install
                    categories: Monitoring, Developer Tools
                    containerImage: quay.io/cryostat/cryostat-operator:2.2.0-dev
                    createdAt: 2021-04-30 00:00:00
                    description: JVM monitoring and profiling tool
                    olm.operatorGroup: operator-sdk-og
                    olm.operatorNamespace: default
                    olm.targetNamespaces: default
                    operatorframework.io/initialization-resource:
                      {
                        "apiVersion": "operator.cryostat.io/v1beta1",
                        "kind": "Cryostat",
                        "metadata": {
                          "name": "cryostat-sample"
                        },
                        "spec": {
                          "enableCertManager": true,
                          "minimal": false,
                          "reportOptions": {
                            "replicas": 0
                          }
                        }
                      }
                    operatorframework.io/properties:
                      {"properties":[{"type":"olm.package","value":{"packageName":"cryostat-operator","version":"2.2.0-dev"}},{"type":"olm.gvk","value":{"group"...
                    operators.operatorframework.io/builder: operator-sdk-v1.4.0+git
                    operators.operatorframework.io/internal-objects: ["flightrecorders.operator.cryostat.io","recordings.operator.cryostat.io"]
                    operators.operatorframework.io/project_layout: go.kubebuilder.io/v3
                    repository: github.com/cryostatio/cryostat-operator
                    support: Cryostat Community
  Service Account:  cryostat-operator-service-account
  Containers:
   manager:
    Image:      quay.io/thvo/cryostat-operator:2.2.0-dev
    Port:       <none>
    Host Port:  <none>
    Command:
      /manager
    Args:
      --leader-elect
    Liveness:   http-get http://:8081/healthz delay=15s timeout=1s period=20s #success=1 #failure=3
    Readiness:  http-get http://:8081/readyz delay=5s timeout=1s period=10s #success=1 #failure=3
    Environment:
      RELATED_IMAGE_CORE:        quay.io/cryostat/cryostat:latest
      RELATED_IMAGE_DATASOURCE:  quay.io/cryostat/jfr-datasource:latest
      RELATED_IMAGE_GRAFANA:     quay.io/cryostat/cryostat-grafana-dashboard:latest
      RELATED_IMAGE_REPORTS:     quay.io/cryostat/cryostat-reports:latest
      WATCH_NAMESPACE:            (v1:metadata.annotations['olm.targetNamespaces'])
      OPERATOR_CONDITION_NAME:   cryostat-operator.v2.2.0-dev
    Mounts:                      <none>
  Volumes:                       <none>
Conditions:
  Type           Status  Reason
  ----           ------  ------
  Available      True    MinimumReplicasAvailable
  Progressing    True    NewReplicaSetAvailable
OldReplicaSets:  <none>
NewReplicaSet:   cryostat-operator-controller-manager-8588cb98b8 (1/1 replicas created)
Events:
  Type    Reason             Age   From                   Message
  ----    ------             ----  ----                   -------
  Normal  ScalingReplicaSet  14m   deployment-controller  Scaled up replica set cryostat-operator-controller-manager-8588cb98b8 to 1
```

Comment (tthvo) [reply bundle/manifests/cryostat-operator.clusterserviceversion.yaml:515]: > I also tried running oc adm top pod cryostat-operator-controller-manager-8588cb98b8-zrgwl but give:

Look like I need to have [metrics-server](https://github.com/kubernetes-sigs/metrics-server) service. I will try this.

Comment (tthvo) [reply bundle/manifests/cryostat-operator.clusterserviceversion.yaml:515]: I think this is the one?
```
NAME                                                    CPU(cores)   MEMORY(bytes)   
cryostat-operator-controller-manager-8588cb98b8-5psst   1m           16Mi
```

Comment (tthvo) [reply bundle/manifests/cryostat-operator.clusterserviceversion.yaml:515]: I am a bit unsure why we set resource request to 64Mi memory and 10m but the operator uses less than that? Was the resource request not applied at all?

Comment (tthvo) [reply bundle/manifests/cryostat-operator.clusterserviceversion.yaml:515]: Any guidance on this @andrewazores @ebaron?

Comment (ebaron) [reply bundle/manifests/cryostat-operator.clusterserviceversion.yaml:515]: We want to ensure we don't set the values too low and then the operator is resource starved or killed.

Comment (andrewazores) [reply bundle/manifests/cryostat-operator.clusterserviceversion.yaml:515]: Going by the limited info we have so far then, I would suggest these as defaults:

```
requests:
  cpu: 100m
  memory: 64Mi
limits:
  cpu: 1000m
  memory: 256Mi
```

The work done shouldn't be very CPU intensive, in fact I would expect it's mostly I/O bound given the amount of k8s API communication it's doing, so 1/10th of a CPU core is probably enough most of the time. I don't think it's doing anything in parallel so more than 1 full core is unnecessary. 64Mi of memory is double the observed steady state so should be plenty including allowance for garbage to accumulate and be collected during active periods, while also being a pretty small request amount in the grand scheme of things. If memory usage gets past 4x that already generous amount then something is probably wrong and the container should probably be restarted.

Comment (ebaron) [reply bundle/manifests/cryostat-operator.clusterserviceversion.yaml:515]: Seems reasonable

Comment (tthvo) [reply bundle/manifests/cryostat-operator.clusterserviceversion.yaml]: Ahh understood! Applied changes :D

Comment (tthvo) [reply bundle/manifests/cryostat-operator.clusterserviceversion.yaml]: >  I wonder if oc adm top pod is behaving like the standard top utility and just printing what's currently in use, not the requests/limits. That would explain why the CPU usage is at 1 millicore, which is effectively zero. I doubt that's the amount assigned, it's just the measured amount at the time you ran the command. If you weren't modifying the Cryostat CR at the time or something then the Operator controller manager was most likely just sitting there idling, waiting for operator-sdk to notify it of some work to do and send it through the reconciliation loop.

I found:
```
$ oc adm top
Show usage statistics of resources on the server

 This command analyzes resources managed by the platform and presents current usage statistics.
Usage:
  oc adm top [flags]

Available Commands:
  images       Show usage statistics for images
  imagestreams Show usage statistics for image streams
  node         Display resource (CPU/memory) usage of nodes
  pod          Display resource (CPU/memory) usage of pods

Use "oc adm <command> --help" for more information about a given command.
Use "oc options" for a list of global command-line options (applies to all commands).
```

Guess I was a bit confused on what it means by `current` :D

**Comment by tthvo** [comment]
I wonder why `describe` on node gives request/limit of the the operator 0?
```
$ oc describe nodes
Namespace                                   Name                                                      CPU Requests  CPU Limits  Memory Requests  Memory Limits  Age
  ---------                                   ----                                                      ------------  ----------  ---------------  -------------  ---
  default                                     cryostat-operator-controller-manager-7bdf6f95b8-hn9m7     0 (0%)        0 (0%)      0 (0%)           0 (0%)         12m
  default                                     cryostat-sample-6f4dc75cc5-2vdwn                          0 (0%)        0 (0%)      0 (0%)           0 (0%)         10m
```

**Review by ebaron — APPROVED** [review]
Looks good to me!

**Comment by andrewazores** [comment]
I'd guess that just means there are no requests/limits. We were never setting one explicitly before and the crc cluster probably doesn't define any limits by default, just so that it's quickest and easiest for developers to run whatever they want on crc for development and not need to worry about tinkering with resources first.

**Comment by ebaron** [comment]
> I wonder why `describe` on node gives request/limit of the operator as 0 (0%)? I thought it would display values as [#423 (comment)](https://github.com/cryostatio/cryostat-operator/pull/423#discussion_r914082667).
> 
> ```
> $ oc describe nodes
> ...(output hidden)
> 
> Namespace                                   Name                                                      CPU Requests  CPU Limits  Memory Requests  Memory Limits  Age
>   ---------                                   ----                                                      ------------  ----------  ---------------  -------------  ---
>   default                                     cryostat-operator-controller-manager-7bdf6f95b8-hn9m7     0 (0%)        0 (0%)      0 (0%)           0 (0%)         12m
>   default                                     cryostat-sample-6f4dc75cc5-2vdwn                          0 (0%)        0 (0%)      0 (0%)           0 (0%)         10m
> 
> ...(output hidden)
> ```

It's showing for me in CRC with your changes applied:
```
  cryostat-operator-system                    cryostat-operator-controller-manager-6f497c598b-w8rc9     100m (2%)     1 (26%)     64Mi (0%)        256Mi (1%)     5s
```

If you're running `make deploy`, this will deploy into the `cryostat-operator-system` namespace by default, fyi.

**Comment by tthvo** [comment]
Oh yes thanks a lot @ebaron. I was using Bundle deployment. But when I switch to manual deployment to `default` namespace. It appears.

```
default     cryostat-operator-controller-manager-57b745f5c8-wkc9v     100m (2%)     1 (26%)     64Mi (0%)        256Mi (1%)     60s
```

Any guidance on why there is such a difference in behavior?

**Comment by ebaron** [comment]
> ```
> 
> ```
> 
> 
>     
>   
> 
> Any guidance on why there is such a difference in behavior when using bundle deployment instead (I was using Bundle deployment above)?

`make deploy_bundle` will install the `quay.io/cryostat/cryostat-operator-bundle:2.2.0-dev` operator bundle by default. You can change this with the `BUNDLE_IMG` variable, but this requires you to build your own bundle and push it to Quay under your namespace. This bundle image includes the Cluster Service Version for the operator, which itself includes the operator deployment manifest. You can see your changes reflected in the CSV when you ran `make bundle`. Running `make deploy_bundle` will not include any local changes you've made.

When you run `make deploy`, any local changes under `config/` are picked up when deploying to the cluster. If you make changes to the Go sources, you'd need to build and push a custom operator image and pass that to `make deploy` with the `OPERATOR_IMG` variable.

Basically YAML definitions in `config/` go in the bundle image, Go code goes in the operator image.

**Comment by tthvo** [comment]
> make deploy_bundle will install the quay.io/cryostat/cryostat-operator-bundle:2.2.0-dev operator bundle by default. You can change this with the BUNDLE_IMG variable, but this requires you to build your own bundle and push it to Quay under your namespace. This bundle image includes the Cluster Service Version for the operator, which itself includes the operator deployment manifest. You can see your changes reflected in the CSV when you ran make bundle. Running make deploy_bundle will not include any local changes you've made.

Ahhh thank you! I think I understand it more now! But when I previously tried:
```
$ export IMAGE_NAMESPACE=quay.io/thvo IMAGE_VERSION=2.2.0-dev
$ make generate manifests manager oci-build bundle bundle-build 
$ podman image prune -f && podman push $IMAGE_NAMESPACE/cryostat-operator:$IMAGE_VERSION && podman push $IMAGE_NAMESPACE/cryostat-operator-bundle:$IMAGE_VERSION
$ make deploy_bundle
```

The images got pushed to my Quay registry. And the bundle images appears to be pulled from there.

```
NAME                                                              READY   STATUS      RESTARTS   AGE
b3d9acc53434464a12c70ad8928b86013dfe1b5cc42962af2958e539ad4vvv8   0/1     Completed   0          6m33s
cryostat-operator-controller-manager-754976d5fd-lrcwk             1/1     Running     0          6m19s
quay-io-thvo-cryostat-operator-bundle-2-2-0-dev                   1/1     Running     0          6m49s
```

But the `oc describe nodes` still shows me 0%.
```
  default                                     cryostat-operator-controller-manager-754976d5fd-lrcwk     0 (0%)        0 (0%)      0 (0%)           0 (0%)         62s
  default                                     quay-io-thvo-cryostat-operator-bundle-2-2-0-dev           0 (0%)        0 (0%)      0 (0%)           0 (0%)         92s
```

I wonder if my steps are correct? Any ideas @ebaron?

**Comment by tthvo** [comment]
> If you make changes to the Go sources, you'd need to build and push a custom operator image and pass that to make deploy with the OPERATOR_IMG variable.

This also means I have to build the image and push it to my own Quay registry right?

**Comment by ebaron** [comment]
> > make deploy_bundle will install the quay.io/cryostat/cryostat-operator-bundle:2.2.0-dev operator bundle by default. You can change this with the BUNDLE_IMG variable, but this requires you to build your own bundle and push it to Quay under your namespace. This bundle image includes the Cluster Service Version for the operator, which itself includes the operator deployment manifest. You can see your changes reflected in the CSV when you ran make bundle. Running make deploy_bundle will not include any local changes you've made.
> 
> Ahhh thank you! I think I understand it more now! But when I previously tried:
> 
> ```
> $ export IMAGE_NAMESPACE=quay.io/thvo IMAGE_VERSION=2.2.0-dev
> $ make generate manifests manager oci-build bundle bundle-build 
> $ podman image prune -f && podman push $IMAGE_NAMESPACE/cryostat-operator:$IMAGE_VERSION && podman push $IMAGE_NAMESPACE/cryostat-operator-bundle:$IMAGE_VERSION
> $ make deploy_bundle
> ```
> 
> The images got pushed to my Quay registry. And the bundle image appears to be pulled from there.
> 
> https://quay.io/repository/thvo/cryostat-operator-bundle?tab=tags
> 
> ```
> NAME                                                              READY   STATUS      RESTARTS   AGE
> b3d9acc53434464a12c70ad8928b86013dfe1b5cc42962af2958e539ad4vvv8   0/1     Completed   0          6m33s
> cryostat-operator-controller-manager-754976d5fd-lrcwk             1/1     Running     0          6m19s
> quay-io-thvo-cryostat-operator-bundle-2-2-0-dev                   1/1     Running     0          6m49s
> ```
> 
> But the `oc describe nodes` still shows me 0%.
> 
> ```
>   default                                     cryostat-operator-controller-manager-754976d5fd-lrcwk     0 (0%)        0 (0%)      0 (0%)           0 (0%)         62s
>   default                                     quay-io-thvo-cryostat-operator-bundle-2-2-0-dev           0 (0%)        0 (0%)      0 (0%)           0 (0%)         92s
> ```
> 
> I wonder if my steps are correct? Any ideas @ebaron?

Your build looks correct to me. I pulled your bundle image and verified that the CSV within it has the resource limits/requests present. Could you check if `oc describe deploy cryostat-operator-controller-manager` shows the correct resource limits and requests?

**Comment by ebaron** [comment]
The CSV shows the expected `resources` property, but the deployments do not. I'm pretty sure we're encountering this bug: https://github.com/operator-framework/operator-lifecycle-manager/issues/1940, which was fixed with https://github.com/operator-framework/api/pull/97. This fix is pulled into Operator SDK 1.6.0, while we're still using 1.5.0.

This bug only affects the `deploy_bundle` install method, so it's annoying for development, but not an issue in production. We really should upgrade the Operator SDK though: https://github.com/cryostatio/cryostat-operator/issues/287.

**Comment by ebaron** [comment]
@tthvo No problem. If you're interested in tackling #287 at some point, I can help you with it.
